File: flutter/platform/lib/src/modules/billing.dart
import '../core/http_client.dart';
import '../core/api_endpoints.dart';
import '../core/result.dart';

/// Billing module — subscription status, plans, entitlements.
class BillingModule {
  const BillingModule();

  /// Get subscription status for current user + app.
  Future<ApiResult<Map<String, dynamic>>> getStatus() =>
      HttpClient.instance.getMap(ApiEndpoints.billingStatus);

  /// Get all entitlements for current user.
  Future<ApiResult<Map<String, dynamic>>> getEntitlements() =>
      HttpClient.instance.getMap(ApiEndpoints.billingEntitlements);

  /// Get available subscription plans.
  Future<ApiResult<Map<String, dynamic>>> getPlans() =>
      HttpClient.instance.getMap(ApiEndpoints.billingPlans);
}

File: web/platform/src/modules/billing.ts
import { http } from "../core/http";
import { Endpoints } from "../core/endpoints";
import type { ApiResult } from "../core/result";

export const billing = {
  getStatus(): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.billingStatus);
  },
  getEntitlements(): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.billingEntitlements);
  },
  // The gateway returns { data: [ ...plans ] } — an array, so getList.
  getPlans(): Promise<ApiResult<unknown[]>> {
    return http.getList(Endpoints.billingPlans);
  },
  /** All of the user's subscriptions (current + past), newest first. */
  listSubscriptions(): Promise<ApiResult<unknown[]>> {
    return http.getList(Endpoints.billingSubscriptions);
  },
  subscribe(planId: string): Promise<ApiResult<Record<string, unknown>>> {
    return http.postMap(Endpoints.billingSubscribe, { planId });
  },
  verifyPayment(body: {
    razorpay_payment_id: string;
    razorpay_subscription_id: string;
    razorpay_signature: string;
  }): Promise<ApiResult<Record<string, unknown>>> {
    return http.postMap(Endpoints.billingVerify, body);
  },
  cancelSubscription(body: {
    razorpaySubscriptionId: string;
    cancelAtCycleEnd?: boolean;
  }): Promise<ApiResult<Record<string, unknown>>> {
    return http.postMap(Endpoints.billingCancel, body);
  },
};
